main.go

```diff
@@ -0,0 +1,376 @@
+			if proto.GetExtension(f.GetOptions(), annotation.E_Public).(bool) || os.Getenv("ALL_PUBLIC") != "" {
+				fieldPath := append(append([]int32{}, msgPath...), 2, int32(j))
+				tField := newField(tFile, msg, fieldPath, f)
+
+				if tField.OneOf != "" {
+					// One-of fields have some extra rules and require some additional
+					// packages.
+					tFile.Imports["net/http"] = true
+					tFile.Imports["reflect"] = true
+					tFile.Imports["strings"] = true
+					tFile.Imports["github.com/istreamlabs/huma"] = true
+					if tMsg.OneOfs[tField.OneOf] == nil {
+						tMsg.OneOfs[tField.OneOf] = []Field{}
+					}
+					tMsg.OneOfs[tField.OneOf] = append(tMsg.OneOfs[tField.OneOf], tField)
+				}
+
+				// Add the new field to the message type.
+				tMsg.Fields = append(tMsg.Fields, tField)
+			}
+		}
+
+		if true || len(tMsg.Fields) > 0 && !tFile.KnownMap[tMsg.Name] {
+			tFile.KnownMap[tMsg.Name] = true
+			tFile.Messages = append(tFile.Messages, tMsg)
+		}
+	}
+}
+
+// buildRegistry makes a map of fully qualified type names to their proto
+// descriptors.
+func buildRegistry(prefix string, messages []*descriptorpb.DescriptorProto, enums []*descriptorpb.EnumDescriptorProto) {
+	for _, msg := range messages {
+		registry[prefix+"."+*msg.Name] = msg
+		buildRegistry(prefix+"."+*msg.Name, msg.NestedType, msg.EnumType)
+	}
+}
+
+func run(input []byte) []byte {
+	// Protoc passes pluginpb.CodeGeneratorRequest in via stdin
+	// marshalled with Protobuf.
+	var req pluginpb.CodeGeneratorRequest
+	proto.Unmarshal(input, &req)
+
+	// Initialise our plugin with default options.
+	opts := protogen.Options{}
+	plugin, err := opts.New(&req)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create a map of files we've been explicitly asked to generate for fast
+	// lookups. The `plugin.Files` used below also contains any imports and we
+	// don't want to generate those in the output.
+	filesToGen := map[string]bool{}
+	for _, file := range plugin.Request.FileToGenerate {
+		filesToGen[file] = true
+	}
+
+	// Generate a type registry map of fully-qualified type names (e.g.
+	// `.google.protobuf.duration`) to the corresponding message or enum
+	// descriptor for lookups later.
+	for _, file := range plugin.Files {
+		buildRegistry("."+*file.Proto.Package, file.Proto.MessageType, file.Proto.EnumType)
+	}
+
+	// Protoc passes a slice of File structs for us to process
+	for _, file := range plugin.Files {
+		if !filesToGen[file.Desc.Path()] {
+			// Skip anything that wasn't explicitly asked for (i.e. built-ins and dependencies).
+			continue
+		}
+
+		tFile := File{
+			Proto:         file.Proto,
+			PackageName:   fmt.Sprintf("%s", file.GoPackageName),
+			Imports:       map[string]bool{string(file.GoImportPath): true},
+			ProtoGoImport: *file.Proto.Options.GoPackage,
+			KnownMap:      map[string]bool{},
+			Messages:      []Message{},
+		}
+
+		// Add all the public types from the file.
+		addEnums(&tFile, "", []int32{5}, file.Proto.EnumType)
```

Comment:
It's because the top-level is a special `FileDescriptorProto` and I've hard-coded the [field numbers](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto#L75-L76). I'll add a comment to make it more clear!